src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
+	}
+	return TRUE;
+}
+
+#if DEBUG
+// Allocate and return a string describing a target
+static char *describe_target (AddReferenceTarget target) {
+	if (target.is_mono_object) {
+		MonoClass *klass = mono.mono_object_get_class (target.obj);
+		return monodroid_strdup_printf ("object of class %s.%s",
+			mono.mono_class_get_namespace (klass),
+			mono.mono_class_get_name (klass));
+	}
+	else
+		return monodroid_strdup_printf ("<temporary object %p>", target.jobj);
+}
+#endif
+
+// Add a reference from one target to another. If the "from" target is a mono_object, it must be a user peer
+static mono_bool
+add_reference (JNIEnv *env, AddReferenceTarget target, AddReferenceTarget reffed_target)
+{
+	MonoJavaGCBridgeInfo *bridge_info = NULL, *reffed_bridge_info = NULL;
+	jobject handle, reffed_handle;
+
+	if (!load_reference_target (target, &bridge_info, &handle))
+		return FALSE;
+
+	if (!load_reference_target (reffed_target, &reffed_bridge_info, &reffed_handle))
+		return FALSE;
+
+	mono_bool success = add_reference_jobject (env, handle, reffed_handle);
+
+	// Flag MonoObjects so they can be cleared in gc_cleanup_after_java_collection.
+	// Java temporaries do not need this because the entire GCUserPeer is discarded.
+	if (success && target.is_mono_object) {
+		int ref_val = 1;
+		mono.mono_field_set_value (target.obj, bridge_info->refs_added, &ref_val);
+	}
+
+#if DEBUG
+	if (gc_spew_enabled) {
+		char *description = describe_target (target),
+			 *reffed_description = describe_target (reffed_target);
+
+		if (success)
+			log_warn (LOG_GC, "Added reference for %s to %s", description, reffed_description);
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+AddReferenceTarget mono_object_target (MonoObject *obj) {
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+AddReferenceTarget jobject_target (jobject jobj) {
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// Extract the root target for an SCC. If the SCC has bridged objects, this is the first object. If not, it's stored in temporary_peers.
+// Must pass in the JNI information necessary to interact with temporary_peers.
+static AddReferenceTarget scc_target (MonoGCBridgeSCC **sccs, int idx, JNIEnv *env, jobject temporary_peers, jmethodID arraylist_get, int *temporary_peer_indices)
+{
+	MonoGCBridgeSCC *scc = sccs [idx];
+	if (scc->num_objs > 0)
+		return mono_object_target (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, arraylist_get, temporary_peer_indices [idx]);
+	return jobject_target (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by scc_target once done with it
+static void release_scc_target (JNIEnv *env, AddReferenceTarget target) {
```

Comment:
`release_scc_target (...)` should go on a newline, and `{` should go on a new line.